src/Agent.Listener/SelfUpdater.cs

```diff
@@ -141,95 +141,150 @@ private async Task DownloadLatestAgent(CancellationToken token)
             IOUtil.DeleteDirectory(latestAgentDirectory, token);
             Directory.CreateDirectory(latestAgentDirectory);
 
+            int agentSuffix = 1;
             string archiveFile;
-            if (_targetPackage.Platform.StartsWith("win"))
-            {
-                archiveFile = Path.Combine(latestAgentDirectory, "agent.zip");
-            }
-            else
-            {
-                archiveFile = Path.Combine(latestAgentDirectory, "agent.tar.gz");
-            }
 
-            Trace.Info($"Save latest agent into {archiveFile}.");
-            try
+            // Download the agent, using multiple attempts in order to be resilient against any networking/CDN issues
+            for (int attempt = 1; attempt <= Constants.AgentDownloadRetryMaxAttempts; attempt++)
             {
-                using (var httpClient = new HttpClient(HostContext.CreateHttpClientHandler()))
+                // Generate an available package name, and do our best effort to clean up stale local zip files
+                while (true)
                 {
-                    //open zip stream in async mode
-                    using (FileStream fs = new FileStream(archiveFile, FileMode.Create, FileAccess.Write, FileShare.None, bufferSize: 4096, useAsync: true))
+                    if (_targetPackage.Platform.StartsWith("win"))
+                    {
+                        archiveFile = Path.Combine(latestAgentDirectory, $"agent{agentSuffix}.zip");
+                    }
+                    else
+                    {
+                        archiveFile = Path.Combine(latestAgentDirectory, $"agent{agentSuffix}.tar.gz");
+                    }
+
+                    try
                     {
-                        using (Stream result = await httpClient.GetStreamAsync(_targetPackage.DownloadUrl))
+                        // delete existing .zip file
+                        if (File.Exists(archiveFile))
                         {
-                            //81920 is the default used by System.IO.Stream.CopyTo and is under the large object heap threshold (85k). 
-                            await result.CopyToAsync(fs, 81920, token);
-                            await fs.FlushAsync(token);
+                            Trace.Verbose("Pre-download: deleting latest agent package zip '{0}'", archiveFile);
+                            IOUtil.DeleteFile(archiveFile);
```

Comment:
i would suggest don't waste time delete file here, delete the entire folder after download finished.
break the loop as soon as we get a valid file name.

Reply:
We can't delete the folder at this specific moment, because it's also used as a staging folder to unzip the agent files before they're finally moved to the bin directory.